src/numeric-types.c

```diff
@@ -93,6 +84,114 @@ numeric_##type##_to_##gtype (numeric_##type num)                     \
     return data.v_##gtype;                                           \
 }
 
+static GArray *TYPE_TABLE = NULL;
+
+/**
+ *
+ */
+void numeric_type_register_static (GType                  type,
+                                   const gchar            *name,
+                                   const NumericTypeInfo *type_info)
+{
+    if (TYPE_TABLE == NULL)
+    {
+        TYPE_TABLE = g_array_new (FALSE, FALSE, sizeof (NumericTypeInfo));
+    }
+
+    /* make sure we don't register the same type twice */
+    gint i;
+    for (i = 0; i < TYPE_TABLE->len; i++)
+    {
+        NumericTypeInfo* ti = &g_array_index (TYPE_TABLE, NumericTypeInfo, i);
+        if (ti->type == type_info->type)
+            return;
+    }
+
+    g_array_append_vals (TYPE_TABLE, type_info, 1);
+}
+
+void
+numeric_type_register_static_simple (GType        type,
+                                     const gchar *name,
+                                     gsize        width,
+                                     gint         byte_order)
+{
+    NumericTypeInfo ti = {
+        .type       = type,
+        .name       = g_strdup (name),
+        .width      = width,
+        .byte_order = byte_order};
+
+    numeric_type_register_static (type, name, &ti);
+}
+
+/*
+ * GLib numerical types already have existing definitions for *_get_type(),
+ * preventing us from registering their numeric_{type}_get_type() counterpart
+ * lazily.
+ */
+static void
+register_glib_numerical_types ()
+{
+    numeric_type_register_static_simple (G_TYPE_CHAR,   "char",   sizeof (gchar),   G_BYTE_ORDER);
+    numeric_type_register_static_simple (G_TYPE_UCHAR,  "uchar",  sizeof (guchar),  G_BYTE_ORDER);
+    numeric_type_register_static_simple (G_TYPE_INT,    "int",    sizeof (gint),    G_BYTE_ORDER);
+    numeric_type_register_static_simple (G_TYPE_UINT,   "uint",   sizeof (guint),   G_BYTE_ORDER);
+    numeric_type_register_static_simple (G_TYPE_LONG,   "long",   sizeof (glong),   G_BYTE_ORDER);
+    numeric_type_register_static_simple (G_TYPE_ULONG,  "ulong",  sizeof (gulong),  G_BYTE_ORDER);
+    numeric_type_register_static_simple (G_TYPE_INT64,  "int64",  sizeof (gint64),  G_BYTE_ORDER);
+    numeric_type_register_static_simple (G_TYPE_UINT64, "uint64", sizeof (guint64), G_BYTE_ORDER);
+    numeric_type_register_static_simple (G_TYPE_FLOAT,  "float",  sizeof (gfloat),  G_BYTE_ORDER);
+    numeric_type_register_static_simple (G_TYPE_DOUBLE, "double", sizeof (gdouble), G_BYTE_ORDER);
+}
```

Comment:
It would be great if we could enumerate these types at runtime so that the we could be forward compatibly if new types are defined in GLib.